zerver/lib/message.py

```diff
@@ -344,3 +347,103 @@ def render_markdown(message, content, realm=None, realm_alert_words=None, messag
         message.is_me_message = Message.is_status_message(content, rendered_content)
 
     return rendered_content
+
+def add_missing_messages(user_profile):
+    # type: (UserProfile) -> None
+    # This list will store all the messages for which we eventually will create
+    # UserMessage table rows by doing a bulk insert.
+    all_messages_to_insert = []  # type: List[UserMessage]
+
+    def store_user_message_to_insert(message):
```

Comment:
Something I didn't notice before is that we really just want to build up a list of message ids, and do all the filtering later on down in the function at the message-id level, not with full objects yet.  And then, only toward the very end, do you need to build up the UserMessage objects, since they're all going to have the same user.